randovania/resolver/resolver.py

```diff
@@ -56,34 +58,39 @@ def _simplify_additional_requirement_set(requirements: RequirementSet,
                           if alternative is not None)
 
 
-def _should_check_if_action_is_safe(state: State, action: ResourceNode, dangerous_resources: frozenset[ResourceInfo],
-                                    ) -> bool:
-    """
-    Determines if we should _check_ if the given action is safe that state
-    :param state:
-    :param action:
-    :return:
-    """
-    if any(resource in dangerous_resources
-           for resource in action.resource_gain_on_collect(state.node_context())):
-        return False
+def _is_action_dangerous(state: State, action: ResourceNode, dangerous_resources: frozenset[ResourceInfo]) -> bool:
+    return any(resource in dangerous_resources
+               for resource, _ in action.resource_gain_on_collect(state.node_context()))
 
-    if isinstance(action, EventNode):
-        return True
 
+def _is_major_or_key_pickup_node(action: ResourceNode, state: State) -> bool:
     if isinstance(action, EventPickupNode):
         pickup_node = action.pickup_node
     else:
         pickup_node = action
 
     if isinstance(pickup_node, PickupNode):
         target = state.patches.pickup_assignment.get(pickup_node.pickup_index)
-        if target is not None and (target.pickup.item_category.is_major or target.pickup.item_category.is_key):
-            return True
-
+        return target is not None and (target.pickup.item_category.is_major or target.pickup.item_category.is_key)
     return False
 
 
+def _should_check_if_action_is_safe(state: State,
+                                    action: ResourceNode,
+                                    dangerous_resources: frozenset[ResourceInfo]
+                                    ) -> bool:
+    """
+    Determines if we should _check_ if the given action is safe that state
+    :param state:
+    :param action:
+    :param dangerous_resources:
+    :return:
+    """
+    return (not _is_action_dangerous(state, action, dangerous_resources)
+            or isinstance(action, EventNode)
```

Comment:
This raises the question. Why are events always safe? There are definitely dangerous events. It's the most common in particular.

Reply:
Oh I see. The old did this. But this new version is logically different!

Reply:
I suppose there are two definitions of safe; one being that the actions not considered dangerous as per the `_is_action_dangerous`  function, and the other includes all the checks that are being done, including the part that involves a `potential_reach`. So previously, due to the bug I'm fixing in 9b416aeea4f7ab29359138d7c3318508200a5ce6, all event actions were considered safe according to the first definition, so the only ones that were considered unsafe are the ones that would be covered by the second definition, which has nothing to do with the event itself, and instead has everything to do with the event being the first action inside a softlock area.

Now that I'm fixing said bug, any actions that are dangerous are correctly defined as unsafe as per the first definition, so by all accounts, not all events are safe.

Side note, I figured that cf5bed2ce45d395e604e10b44afc70d232c7b4b2 provides a more idiomatic way to resolve the bug.

Reply:
Oh, so the old if was completely broken. And checking back, since always that check was added. Neato.
Also I did some testing and unpacking the tuple is faster than indexing it. That was very surprising.

There's probably a better name for `_is_action_dangerous` but that's more picky-ness.